### ISSUE TITLE
reg_max, request for help

### ISSUE THREAD

**Issue by Ange1ika** [ultralytics/ultralytics#20597]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I want to ask: why I cant use ddp (when I train with yolo.train(device=[0,1,2])), if I change reg_max = 1, and off the dfl module? Maybe I wrong change reg_max? My changers on git@github.com:Ange1ika/ultralytics_fine_tune-reg_max.git. When I train with my version, all weights and train are holding only first gpu, despite adding arg device=[0,1,2]. 

when I try train with yours repo and without changes reg - picture 1

![Image](https://github.com/user-attachments/assets/e60d06cb-d78c-4f15-87e0-957f26d079a2)

on the other hand: when I use my approach

![Image](https://github.com/user-attachments/assets/603ce56a-a876-4463-80bd-adf072ffae91)

Can you help, please: how right change reg_max and how fine-tune (for instance segmentation task), on our custom dataset, where I have very small and very large masks

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Ange1ika, thank you for reaching out and for providing detailed context with images and your custom repo link! 🚀 This is an automated response to assist you quickly—an Ultralytics engineer will also review your question and provide further help soon.

We recommend checking out the [Docs](https://docs.ultralytics.com/) for guidance, especially the [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples, which may address some common scenarios.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE). This helps us debug your issue efficiently.

For custom training ❓ questions, it helps to include as much info as possible—dataset image examples, training logs, and confirmation that you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community to discuss your setup! For real-time chat, join us on [Discord](https://discord.com/invite/ultralytics) 🎧. For more detailed threads, visit [Discourse](https://community.ultralytics.com/), or share insights on our [Subreddit](https://reddit.com/r/Ultralytics).

## Upgrade

Make sure you’re running the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your patience!

**Comment by Y-T-G** [comment]
Did you install your local repo with `pip install .` (dot at end)?

**Comment by Ange1ika** [comment]
> Did you install your local repo with `pip install .` (dot at end)?

Yes, I do

**Comment by Ange1ika** [comment]
I changed ultralytics for training with arbitrary reg_max. It is important to change the reg_max parameter in the scripts train_v2, ~/ultralytics/ultralytics/nn/modules/head.py , ~/ultralytics/ultralytics/ultralytics/utils/tal.py. The custom configuration for yolov11m-seg is also used(the only change is in the last line - [[16, 19, 22], 1, Segment, [nc, 128, 256]] # Detect(P3, P4, P5) For example, if reg_max=128 ) The parameter is responsible for the dfl branch in the segmentation head. Details:
https://zhuanlan.zhihu.com/p/13303751794#:~:text=%23%20post%2Dprocess%0A%20%20%20%20bboxes%2C%20scores%2C%20labels%20%3D%20post_process_det(%0A%20%20%20%20%20%20%20%20score_preds%2C%20xyxy_preds%2C%20conf_thresh%3D0.2%2C%20nms_thresh%3D0.5%2C%20num_classes%3D80)

**Comment by Ange1ika** [comment]
> Can you provide the output after running this command in terminal: `yolo checks`?

Yea, I can. I'll do it tomorrow, there's no way to show it now.

**Comment by Ange1ika** [comment]
Thank you for your promptness

**Comment by Ange1ika** [comment]
Thank you so much! I'll try. I used this code. I use Segmentor and a little trick. 

https://github.com/Ange1ika/ultralytics_fine_tune-reg_max/blob/main/train_v2.py#L61C1-L80C6

**Comment by glenn-jocher** [comment]
Looking at your train_v2.py code, I see you're using a custom `Segmentor` class. The issue with multi-GPU training after modifying `reg_max` likely stems from inconsistencies in how the model is defined and distributed across GPUs.

When making architectural changes like setting `reg_max=1`, you need to ensure:

1. All model components are properly initialized before DDP wrapping happens
2. The loss function in utils/loss.py properly handles your modified architecture 
3. The model outputs maintain consistent shapes across all components

For your segmentation task with very small and large masks, instead of disabling DFL completely, you might consider:

- Using a higher `reg_max` value but adjusting the loss weights
- Implementing a custom segmentation head that better handles your dataset's characteristics
- Using mosaic/mixup augmentations to better handle varying mask scales

To debug your current issue, try adding this before your model.train() call:
```python
model.args.device = [0,1,2]  # Explicitly set devices
print(f"Device configuration: {model.args.device}")
```

This will help verify if your device settings are being properly passed to the trainer.

**Comment by Ange1ika** [comment]
> Looking at your train_v2.py code, I see you're using a custom `Segmentor` class. The issue with multi-GPU training after modifying `reg_max` likely stems from inconsistencies in how the model is defined and distributed across GPUs.
> 
> When making architectural changes like setting `reg_max=1`, you need to ensure:
> 
> 1. All model components are properly initialized before DDP wrapping happens
> 2. The loss function in utils/loss.py properly handles your modified architecture 
> 3. The model outputs maintain consistent shapes across all components
> 
> For your segmentation task with very small and large masks, instead of disabling DFL completely, you might consider:
> 
> - Using a higher `reg_max` value but adjusting the loss weights
> - Implementing a custom segmentation head that better handles your dataset's characteristics
> - Using mosaic/mixup augmentations to better handle varying mask scales
> 
> To debug your current issue, try adding this before your model.train() call:
> ```python
> model.args.device = [0,1,2]  # Explicitly set devices
> print(f"Device configuration: {model.args.device}")
> ```
> 
> This will help verify if your device settings are being properly passed to the trainer.

Okeeey! Thank you very much, I try it. The big reg_max (32 and 64) improve large masks, but so, I have lost a small mask-segmentation. If I check 128 reg_max - the model has broken and as a result I get zero segmentation objects. 

I am very grateful for feedback and on this week I try all your suggestions

**Comment by glenn-jocher** [comment]
Thank you for the update on your experiments. It's interesting to see that higher `reg_max` values (32, 64) improved large mask detection but compromised small mask performance, while 128 broke the model completely.

For handling both small and large masks effectively, you might want to try:

1. A moderate `reg_max` value (like 24 or 32) combined with data augmentation specifically targeting small objects (random scaling, cropping)

2. Consider modifying the segmentation loss function to give more weight to small masks:
```python
# Example approach in calculate_segmentation_loss function
mask_areas = (masks > 0).sum(dim=(1,2))
small_mask_weight = torch.exp(-mask_areas / area_threshold)
weighted_loss = mask_loss * (1 + small_mask_weight)
```

3. For multi-GPU training, ensure your model architecture modifications are applied consistently before DDP initialization by adding explicit model preparation:
```python
# Before training
model.model[-1].reg_max = your_reg_max_value
# Update any related parameters in loss functions
model.loss = YourModifiedLoss(reg_max=your_reg_max_value)
```

Let me know how these suggestions work for your specific dataset characteristics.

**Comment by Ange1ika** [comment]
Hello! Now I am working with this task and I noticed, that in the tal.py the reg_max is given separatly, without self.reg_max from model. Why and Can I ask you about tal.py, what main goal this script?

![Image](https://github.com/user-attachments/assets/0e0dd13d-9a9b-4ccf-90e8-113e7a1095fc)

**Comment by Ange1ika** [comment]
![Image](https://github.com/user-attachments/assets/49fe6e5d-6a58-4ce9-8f99-9645ea341785)


Managed to start it with Multi-GPU, but because weights are loaded unevenly at the very beginning, I have CUDA_OUT_of memory. 
It doesn't matter anymore, no big deal, I'll try to train on a smaller dataset

**Comment by glenn-jocher** [comment]
Thanks for sharing your progress! Regarding `tal.py`, this file implements the Task-Aligned assignment algorithm that matches predictions with ground truth during training. The `reg_max` is defined separately there to make the assigner independent from the model - it needs to decode boxes consistently with how they're encoded in the detection head.

For your CUDA memory issues with multi-GPU training, this is likely happening because the model's weights are being distributed unevenly. You could try:

1. Reducing batch size per GPU with something like `batch=8`
2. Using gradient accumulation by adding `accumulate=2` to spread the memory load
3. Setting `--autobatch` flag to automatically find the optimal batch size

For handling both small and large masks effectively, a balanced approach might be setting `reg_max` to 16-24 combined with custom augmentations for small objects. The standard YOLO model's `reg_max=16` is a carefully tuned default that balances performance across different object scales.
